### PR TITLE
Add option `cwd` to running tasks

### PR DESCRIPTION
**Pull request by Diogo-Rossi**
TL;DR:

Implements #81 

---

Following the suggestion of @illBeRoy:
> 1. Be able to define CWD globally via:
> 
> ```toml
> [tool.taskipy.settings]
> cwd = "."
> ```
> 
> 2. Be able to define it per command:
> 
> ```toml
> [tool.taskipy.tasks]
> lint = { cmd = "pylint", cwd = '.' }
> ```
> 
> Where:
> 
> 1. If CWD is defined both globally and per a specific task, the one on the task takes precedence
> 2. CWD allows for relative paths

Tested running tasks from folder `./vscode` with the following tasks:

```toml
[tool.taskipy.tasks]
pwdroot = { cmd = "python -c \"import os; print(os.getcwd())\"" , cwd = "."}
pwdtest = { cmd = "python -c \"import os; print(os.getcwd())\"" , cwd = "./tests"}
pwddocs = { cmd = "python -c \"import os; print(os.getcwd())\"" , cwd = "./docs"}
pwderro = { cmd = "python -c \"import os; print(os.getcwd())\"" , cwd = "./non_existent_folder"}
pwdabs = { cmd = "python -c \"import os; print(os.getcwd())\"" , cwd = "C:/Users"}
```

Which results are:
```
❯ task pwdroot
C:\Users\user\Repos\taskipy

❯ task pwdtest
C:\Users\user\Repos\taskipy\tests

❯ task pwddocs
C:\Users\user\Repos\taskipy\docs

❯ task pwderro
C:\Users\user\Repos\taskipy\.vscode

❯ task pwdabs
C:\Users
```

### CODE REVIEW

File: taskipy/task.py [hunk 1]
```diff
@@ -48,7 +53,19 @@ def __extract_task_command(self, task_toml_contents: object) -> str:
...
-        raise MalformedTaskError(self.__task_name, 'tasks must be strings, or dicts that contain { cmd, help, use_vars }')
+        raise MalformedTaskError(self.__task_name, 'tasks must be strings, or dicts that contain { cmd, cwd, help, use_vars }')
+
+    def __extract_task_workdir(self, task_toml_contents: object) -> Optional[str]:
+        if isinstance(task_toml_contents, str):
+            return None
+
+        if isinstance(task_toml_contents, dict):
+            try:
+                return task_toml_contents['cwd']
```
Comment (illBeRoy) [inline taskipy/task.py:64]: I was thinking: maybe we should also assert that `cwd` is a string, and throw if it's not. wdyt?

File: taskipy/task_runner.py [hunk 2]
```diff
@@ -24,8 +24,8 @@
 class TaskRunner:
     def __init__(self, cwd: Union[str, Path]):
         working_dir = cwd if isinstance(cwd, Path) else Path(cwd)
```
Comment (illBeRoy) [inline taskipy/task_runner.py:26]: I think that having two variables called `working_dir` and `self.__working_dir` (line 28) might be confusing, as they are not used for the same purpose.
Perhaps we can rename this variable into `cwd_as_path` to emphasize what it's really used for?

**Review by illBeRoy — CHANGES_REQUESTED** [review]
All in all, this looks really nice. Thanks for the contribution!



In addition to the code requests, can you please add tests to tests/test_taskipy.py ?
You can add a new class `SetCWDTestCase`, with the following tests:
1. With cwd under `tool.taskipy.settings.cwd`
2. With cwd under a specific task
3. With cwd in both settings + the specific task, make sure task takes precedence
4. Make sure that relative CWD is calculated relative to project regardless of where the command was called

Thank you!

File: taskipy/task.py (hunk 1 above)
Comment (Diogo-Rossi) [reply taskipy/task.py:64]: I think it will be good!

File: taskipy/task_runner.py (hunk 2 above)
Comment (Diogo-Rossi) [reply taskipy/task_runner.py:26]: Ok!

**Comment by Diogo-Rossi** [comment]
I'm still writing tests.
They will be ready very soon!

**Comment by Diogo-Rossi** [comment]
@illBeRoy 
Done!

File: taskipy/task.py (hunk 1 above)
Comment (Diogo-Rossi) [reply taskipy/task.py]: Done on eff3a19e36a72d4a7644681c0cb61dc759aa499c

File: taskipy/task_runner.py (hunk 2 above)
Comment (Diogo-Rossi) [reply taskipy/task_runner.py]: Done on 7822e377e2aa5a65f533779770606c309760913c

**Comment by illBeRoy** [comment]
seems like some tests are failing. can you please take a look?

**Comment by Diogo-Rossi** [comment]
> seems like some tests are failing. can you please take a look?
Ok!

**Comment by Diogo-Rossi** [comment]
> seems like some tests are failing. can you please take a look?

Sorry, but I took a look in the tests: I only introduced 3 tests in a test case, and they seem to be passing:

<details>

image

image

image

image

image

</details>

**Comment by Diogo-Rossi** [comment]
Oh, I just saw the failing test: it is related to the error message, which was changed.

```python
self.assertSubstr('tasks must be strings, or dicts that contain { cmd, cwd, help, use_vars }', stdout)
```

I'll fix that.

**Comment by illBeRoy** [comment]
It seems like lint check is failing now. It's mostly over whitespaces so seems like this could be easily fixed.
Once you're done, you can try running `poetry run task test` locally and see that everything works as expected :)

**Comment by Diogo-Rossi** [comment]
> It seems like lint check is failing now. It's mostly over whitespaces so seems like this could be easily fixed. Once you're done, you can try running `poetry run task test` locally and see that everything works as expected :)

I'm getting this error:

```python
ModuleNotFoundError: No module named 'tests/test_*'
```

**Comment by Diogo-Rossi** [comment]
Ok, I just fixed the spaces issue and run local tests manually with `task lint_pylint` and `task lint_mypy`, the results are bellow.

OBS:
I could not use `task lint` because it was not recognized by my shell.
(Maybe because I'm on Windows, with a PowerShell terminal?).

Anyway, the `task test` command (or `poetry run task test` command) also did not worked, as said before. It is giving error when importing the module:

```PowerShell
❯ task test
tests/test_* (unittest.loader._FailedTest.tests/test_*) ... ERROR

======================================================================
ERROR: tests/test_* (unittest.loader._FailedTest.tests/test_*)
----------------------------------------------------------------------
ImportError: Failed to import test module: tests/test_*
Traceback (most recent call last):
  File "[...]\Lib\unittest\loader.py", line 137, in loadTestsFromName
    module = __import__(module_name)
             ^^^^^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'tests/test_*'
```

Is it correct to run the tests with this pattern `tests/test_*`? I did not find it on [the docs](https://docs.python.org/3/library/unittest.html).

<details>
<summary>pylint results</summary>

```PowerShell
❯ task lint_pylint
************* Module D:\Repos\taskipy\.pylintrc
.pylintrc:1:0: E0015: Unrecognized option found: missing-module-docstring (unrecognized-option)
.pylintrc:1:0: R0022: Useless option value for '--disable', 'R0201' was moved to an optional extension, see https://pylint.readthedocs.io/en/latest/whatsnew/2/2.14/summary.html#removed-checkers. (useless-option-value)
```
</details>

<details>
<summary>mypy results</summary>
```PowerShell
❯ task lint_mypy
tests\test_taskipy.py:24: note: By default the bodies of untyped functions are not checked, consider using --check-untyped-defs  [annotation-unchecked]
Success: no issues found in 18 source files
```
</details>

**Comment by illBeRoy** [comment]
> Anyway, the task test command (or poetry run task test command) also did not worked, as said before. It is giving error when importing the module:

Interesting. Maybe this is because in unix systems, globs are resolved before they're passed into the receiving process and on Windows this is not the case.

I'll take a look about best practices with running entire test directories using `unittest`.

Anyways, seem like the CI is passing! Great job mate :) I'll try to review today \ tomorrow at the latest.

Thanks again for taking the time to contribute!

**Comment by illBeRoy** [comment]
Hey! Just letting you know that I haven't forgotten about this PR. Just swamped with work. Will try to get it done by Sunday :)

**Review by illBeRoy — APPROVED** [review]
Looks great! Thank you so much for your contribution.

**Comment by stevenschmatz** [comment]
Hey team! I would love to use this feature. However I don't think a release has been published with this yet.

**Comment by stevenschmatz** [comment]
(For now I'm just using the package from git directly):

```shell
uv add git+https://github.com/taskipy/taskipy
```

**Pull request merged**